sim/message_queue.go

```diff
@@ -1,49 +1,111 @@
 )
 
+var _ heap.Interface = (*messageQueue)(nil)
+
+// messageQueue is a priority queue that implements heap.Interface and holds
+// simulation messages in flight prioritised by their delivery time in ascending
+// order.
+type messageQueue struct {
+	mailbox []*messageInFlight
+}
+
 type messageInFlight struct {
 	source    gpbft.ActorID // ID of the sender
 	dest      gpbft.ActorID // ID of the receiver
-	payload   interface{}   // Message body
+	payload   any           // Message body
 	deliverAt time.Time     // Timestamp at which to deliver the message
+
+	index int // Index in the heap used internally by the heap implementation
 }
 
-// A queue of directed messages, maintained as an ordered list.
-type messageQueue []messageInFlight
+func newMessagePriorityQueue() *messageQueue {
+	var mpq messageQueue
+	heap.Init(&mpq)
+	return &mpq
+}
+
+// Len returns the number of messages that are currently in-flight.
+func (pq *messageQueue) Len() int { return len(pq.mailbox) }
+
+// Less determines whether message in-flight at index i should be prioritised
+// before the message in-flight at index j. The message deliverAt filed is used
+// to determine this, where the earlier the deliverAt the higher priority the
+// message, i.e. messages are sorted in ascending order of their deliverAt.
+//
+// This function is part of heap.Interface and must not be called externally.
+func (pq *messageQueue) Less(i, j int) bool {
+	// We want Pop to give us the earliest delivery time, so we use Less to sort by
+	// deliverAt in ascending order.
+	return pq.mailbox[i].deliverAt.Before(pq.mailbox[j].deliverAt)
+}
 
-func (h *messageQueue) Insert(x messageInFlight) {
-	// Insert the new message after any messages with a sooner or equal deliverAt.
-	i := sort.Search(len(*h), func(i int) bool {
-		ith := (*h)[i].deliverAt
-		return ith.After(x.deliverAt)
-	})
-	*h = append(*h, messageInFlight{})
-	copy((*h)[i+1:], (*h)[i:])
-	(*h)[i] = x
+// Swap swaps messages in-flight at index i with the one at index j.
+//
+// This function is part of heap.Interface and must not be called externally.
+func (pq *messageQueue) Swap(i, j int) {
+	pq.mailbox[i], pq.mailbox[j] = pq.mailbox[j], pq.mailbox[i]
+	pq.mailbox[i].index = i
+	pq.mailbox[j].index = j
 }
 
-// Removes an entry from the queue
-func (h *messageQueue) Remove(i int) messageInFlight {
-	v := (*h)[i]
-	copy((*h)[i:], (*h)[i+1:])
-	*h = (*h)[:len(*h)-1]
-	return v
+// Push adds an element to this queue.
+//
+// This function is part of heap.Interface and must not be called externally.
+// See: Insert.
+func (pq *messageQueue) Push(x any) {
+	n := len(pq.mailbox)
+	item := x.(*messageInFlight)
+	item.index = n
+	pq.mailbox = append(pq.mailbox, item)
+}
+
+// Pop removes and returns the highest priority message from the queue.
+//
+// This function is part of heap.Interface and must not be called externally.
+// See: Remove.
+func (pq *messageQueue) Pop() any {
+	old := pq.mailbox
+	n := len(old)
+	item := old[n-1]
```

Comment:
This looks like it's removing the lowest priority (highest deliverAt) item, if they are in ascending order. Which I don't think is what we want here.

This message queue is lacking tests (my fault).

Reply:
The implementation of `Pop` is correct, i.e. adheres to what `heap.Interface` demands. It's my documentation that's misleading; apologies.

Unit tests are added [here](https://github.com/filecoin-project/go-f3/pull/221) along with documentation correction.